Add ClusterMembership Persistence Capabilities

shawnhathaway opened the pull request:
This is related to https://github.com/temporalio/temporal/issues/83.

This adds three apis to the ClusterMetadataManager across mysql, postgresql and Cassandra:

```
- GetActiveClusterMembers(request *GetActiveClusterMembersRequest (*GetActiveClusterMembersResponse, error)
- UpsertClusterMembership(request *UpsertClusterMembershipRequest) error
- PruneClusterMembership(request *PruneClusterMembershipRequest) error
```

This paves the way for a subsequent change to persist and coordinate ringpop membership through this layer.

samarabbas commented on common/persistence/cassandra/cassandraClusterMetadata.go:
```diff
@@ -33,18 +37,29 @@ import (
 	immutableEncodingFieldName = immutablePayloadFieldName + `_encoding`
+
+	// ****** CLUSTER_MEMBERSHIP TABLE ******
+	templateUpsertActiveClusterMembership = `UPDATE
```
You may want to use INSERT here otherwise we may leak keys.

samarabbas commented on common/persistence/cassandra/cassandraClusterMetadata.go:
```diff
@@ -148,3 +163,49 @@ func (m *cassandraClusterMetadata) GetImmutableClusterMetadata() (*p.InternalGet
+func (m *cassandraClusterMetadata) GetActiveClusterMembers(request *p.GetActiveClusterMembersRequest) (*p.GetActiveClusterMembersResponse, error) {
+	query := m.session.Query(templateGetActiveClusterMembership, time.Now().UTC().Add(-request.LastHeartbeatWithin))
+
+	// No paging enabled due to TTL on table
```
I'm confused about this comment here.  How is TTL related to pagination?  Actually it would be good if we support pagination for the very beginning.

samarabbas commented on common/persistence/dataInterfaces.go:
```diff
@@ -1449,6 +1449,37 @@ type (
 
+	// GetActiveClusterMembersRequest is the response to GetActiveClusterMembers
+	GetActiveClusterMembersRequest struct {
+		LastHeartbeatWithin time.Duration
```
I thought we already have expiry, what's the point of taking in this parameter?

mfateev replied:
We need a way to list all members including timed out ones. 
So I would rename this method to GetClusterMembers(LastHeartbeatWithin)

shawnhathaway replied:
Discussed in person that expiry is for long term expiration out of table (say 48 hours), but services will apply a much shorter expiration (say 10 minutes) from the read side on LastHeartbeat.

samarabbas commented on common/persistence/dataInterfaces.go:
```diff
@@ -1449,6 +1449,37 @@ type (
+		HostID         uuid.UUID
+		RPCAddress     string
+		SessionStarted time.Time
+		LastHeartbeat  time.Time
```
Let's add expiry as another field.

mfateev replied:
What about role?

mfateev commented on schema/cassandra/cadence/schema.cql:
```diff
@@ -475,3 +475,14 @@ CREATE TABLE queue (
   };
+
+CREATE TABLE cluster_membership
+(
```
partition key is missing

shawnhathaway replied:
I have been deferring partitionKey until after this and the cassandra frozen blob proto rework as I need to dig a bit deeper into the required changes on our side for partitioning and if there is anything additional for supporting vitess. 

When implementing partition keys for mysql is it correct that we would want to avoid spanning multiple partitions in queries? Assuming, so we'd want to pin all records to the same partition similarly to Cassandra, however, does the absence of a partition key not give the same behavior?

mfateev replied:
Yes, we want all the records to live in a single partition. In case of Cassandra absence of partition key is anti-pattern: https://dba.stackexchange.com/questions/157537/querying-cassandra-without-a-partition-key

shawnhathaway replied:
Ah yes, had good convo with Samar on this, thanks you. I meant to reply to one of the .sql files in reference to postgres/mysql behavior without a PK.

mfateev commented on schema/mysql/v57/cadence/schema.sql:
```diff
@@ -254,10 +254,19 @@ CREATE TABLE queue_metadata (
+);
+
+CREATE TABLE cluster_membership
+(
```
partition key is missing

mfateev commented on schema/postgres/cadence/schema.sql:
```diff
@@ -259,4 +259,14 @@ CREATE TABLE cluster_metadata (
+);
+
+CREATE TABLE cluster_membership
+(
```
partition key is missing

shawnhathaway commented:
This last set of commits addresses all feedback other than the paging impl (coming shortly) and PartitionKeys for Mysql (deferring based on current understanding pending convo)

samarabbas commented on common/persistence/dataInterfaces.go:
```diff
@@ -1449,30 +1450,40 @@ type (
+		RPCAddressEquals    net.IP
+		HostIDEquals        uuid.UUID
+		RoleEquals          ServiceType
+		PageSize			int
```
fix formatting.

samarabbas commented on common/persistence/dataInterfaces.go:
```diff
@@ -1449,30 +1450,40 @@ type (
+		HostIDEquals        uuid.UUID
+		RoleEquals          ServiceType
+		PageSize			int
+		NextPageToken		[]byte
```
Maybe remove for now.  You can add it back when you implement pagination.

shawnhathaway replied:
Whoops this wasnt supposed to make it in :)

samarabbas commented on common/persistence/dataInterfaces.go:
```diff
@@ -2538,3 +2549,13 @@ func NewGetReplicationTasksFromDLQRequest(
+type ServiceType int
+
+const (
+	Unknown ServiceType = iota
```
Maybe call Unknown as All?  As the behavior is to return all hosts irrespective of the role.

shawnhathaway replied:
Moved to All, however I think we may want to have a default of None/Unknown = defaultValue for any type that we persist (may just want to do it everywhere). This is a practice is some circles but lets discuss.
